server/conn.go

```diff
@@ -1254,55 +1258,83 @@ func (cc *clientConn) handleIndexAdvise(ctx context.Context, indexAdviseInfo *ex
 // There is a special query `load data` that does not return result, which is handled differently.
 // Query `load stats` does not return result either.
 func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
-	rss, err := cc.ctx.Execute(ctx, sql)
+	stmts, err := cc.ctx.Parse(ctx, sql)
 	if err != nil {
 		metrics.ExecuteErrorCounter.WithLabelValues(metrics.ExecuteErrorToLabel(err)).Inc()
 		return err
 	}
-	status := atomic.LoadInt32(&cc.status)
-	if rss != nil && (status == connStatusShutdown || status == connStatusWaitShutdown) {
-		for _, rs := range rss {
-			terror.Call(rs.Close)
-		}
-		return executor.ErrQueryInterrupted
-	}
-	if rss != nil {
-		if len(rss) == 1 {
-			err = cc.writeResultset(ctx, rss[0], false, 0, 0)
-		} else {
-			err = cc.writeMultiResultset(ctx, rss, false)
-		}
-	} else {
-		loadDataInfo := cc.ctx.Value(executor.LoadDataVarKey)
-		if loadDataInfo != nil {
-			defer cc.ctx.SetValue(executor.LoadDataVarKey, nil)
-			if err = cc.handleLoadData(ctx, loadDataInfo.(*executor.LoadDataInfo)); err != nil {
-				return err
+
+	var rs ResultSet
+	for i, stmt := range stmts {
+		rs, err = cc.ctx.ExecuteStmt(ctx, stmt)
+		if err != nil {
+			if rs != nil {
+				terror.Call(rs.Close)
 			}
+			break
 		}
 
-		loadStats := cc.ctx.Value(executor.LoadStatsVarKey)
-		if loadStats != nil {
-			defer cc.ctx.SetValue(executor.LoadStatsVarKey, nil)
-			if err = cc.handleLoadStats(ctx, loadStats.(*executor.LoadStatsInfo)); err != nil {
-				return err
-			}
+		status := cc.ctx.Status()
+		if i != len(stmts)-1 {
+			status |= mysql.ServerMoreResultsExists
 		}
 
-		indexAdvise := cc.ctx.Value(executor.IndexAdviseVarKey)
-		if indexAdvise != nil {
-			defer cc.ctx.SetValue(executor.IndexAdviseVarKey, nil)
-			err = cc.handleIndexAdvise(ctx, indexAdvise.(*executor.IndexAdviseInfo))
+		if rs != nil {
+			connStatus := atomic.LoadInt32(&cc.status)
+			if connStatus == connStatusShutdown || connStatus == connStatusWaitShutdown {
+				terror.Call(rs.Close)
+				err = executor.ErrQueryInterrupted
+				break
+			}
+
```

Comment:
Should we handle `MultiQueryNoDelayResult` here?

Reply:
There should be no `MultiQueryNoDelayResult` here, because we do not execute multiple query in one Execute any more.

I'll clean up the unused struct and related code in the following PRs.